dataset link: https://drive.google.com/file/d/1uXmuA1GfHYmF9AaVRoNoeKt4W1MPL_Gn/view?usp=sharing

# **Task 1 Identify Features + Task 3 Feature Engineering**

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv('vehicles.csv')

print(df.shape)


(423857, 25)


In [2]:
df.head(10)

id                                                url  \
0  7184791621  https://duluth.craigslist.org/ctd/d/duluth-200...   
1  7184773187  https://duluth.craigslist.org/cto/d/saginaw-20...   
2  7193375964  https://newhaven.craigslist.org/cto/d/stratfor...   
3  7195108810  https://albuquerque.craigslist.org/cto/d/albuq...   
4  7184712241  https://duluth.craigslist.org/ctd/d/rush-city-...   
5  7184711787  https://duluth.craigslist.org/ctd/d/rush-city-...   
6  7184710887  https://duluth.craigslist.org/ctd/d/rush-city-...   
7  7184710023  https://duluth.craigslist.org/ctd/d/rush-city-...   
8  7184709156  https://duluth.craigslist.org/ctd/d/rush-city-...   
9  7184708752  https://duluth.craigslist.org/ctd/d/rush-city-...   

              region                          region_url  price    year  \
0  duluth / superior       https://duluth.craigslist.org   6995  2000.0   
1  duluth / superior       https://duluth.craigslist.org   8750  2013.0   
2          new haven     https://newhaven.craigslist.org  10900  2013.0   
3        albuquerque  https://albuquerque.craigslist.org  12500  2003.0   
4  duluth / superior       https://duluth.craigslist.org  16995  2007.0   
5  duluth / superior       https://duluth.craigslist.org  13995  2012.0   
6  duluth / superior       https://duluth.craigslist.org   7995  2010.0   
7  duluth / superior       https://duluth.craigslist.org   8995  2011.0   
8  duluth / superior       https://duluth.craigslist.org  10995  2014.0   
9  duluth / superior       https://duluth.craigslist.org  12995  2004.0   

  manufacturer                  model  condition    cylinders  ... drive  \
0          gmc        new sierra 1500  excellent  8 cylinders  ...   4wd   
1      hyundai                 sonata  excellent  4 cylinders  ...   fwd   
2       toyota                  prius       good  4 cylinders  ...   fwd   
3   mitsubishi                 lancer       good  4 cylinders  ...   4wd   
4          gmc  sierra classic 2500hd       good  8 cylinders  ...   4wd   
5         ford                  f-150       good  6 cylinders  ...   4wd   
6    chevrolet                equinox       good  4 cylinders  ...   4wd   
7    chevrolet               traverse       good  6 cylinders  ...   4wd   
8         ford               explorer       good  6 cylinders  ...   4wd   
9          ram                   2500       good  6 cylinders  ...   4wd   

        size   type paint_color  \
0        NaN    NaN         red   
1        NaN    NaN        grey   
2        NaN    NaN        blue   
3   mid-size  sedan        grey   
4  full-size  truck       white   
5  full-size  truck        grey   
6  full-size    SUV        grey   
7  full-size    SUV       white   
8  full-size    SUV       white   
9  full-size  truck      custom   

                                           image_url  \
0  https://images.craigslist.org/00n0n_f06ykBMcdh...   
1  https://images.craigslist.org/00d0d_kgZ6xoeRw2...   
2  https://images.craigslist.org/00d0d_3sHGxPbY2O...   
3  https://images.craigslist.org/00m0m_4a8Pb6JbMG...   
4  https://images.craigslist.org/01414_g093aPtSMW...   
5  https://images.craigslist.org/00W0W_io2cwEVtyh...   
6  https://images.craigslist.org/00404_j5GItixnjX...   
7  https://images.craigslist.org/00T0T_bwYCuKicjX...   
8  https://images.craigslist.org/00O0O_j7JSL1b8Qr...   
9  https://images.craigslist.org/00w0w_7Igao5t6ZX...   

                                         description county state      lat  \
0  2000 *** GMC New Sierra 1500 Ext Cab 157.5  WB...    NaN    mn  46.8433   
1  For Sale:  2013 Hyundai Sonata GLS - $8750.  O...    NaN    mn  46.9074   
2  2013 Prius V Model Two. One owner—must sell my...    NaN    ct  41.1770   
3  2003 Mitsubishi Lancer Evolution, silver.  Abo...    NaN    nm  35.1868   
4  **Bad Credit, No Credit... No Problem!**2007 G...    NaN    mn  45.6836   
5  **Bad Credit, No Credit... No Problem!**2012 F...    NaN    mn  45.6836   
6  **Bad Credit, No Credit... No Problem!**2010 C.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 423857 entries, 0 to 423856
Data columns (total 25 columns):
id              423857 non-null int64
url             423857 non-null object
region          423857 non-null object
region_url      423857 non-null object
price           423857 non-null int64
year            328743 non-null float64
manufacturer    313242 non-null object
model           325384 non-null object
condition       176719 non-null object
cylinders       197679 non-null object
fuel            327214 non-null object
odometer        270585 non-null float64
title_status    327759 non-null object
transmission    328065 non-null object
vin             184420 non-null object
drive           231119 non-null object
size            102627 non-null object
type            241157 non-null object
paint_color     222203 non-null object
image_url       329661 non-null object
description     329654 non-null object
county          0 non-null float64
state           423857 non-null obj

# Cleaning data
เช็คดูค่า null ใน dataframe

In [4]:
df.isnull().sum()

id                   0
url                  0
region               0
region_url           0
price                0
year             95114
manufacturer    110615
model            98473
condition       247138
cylinders       226178
fuel             96643
odometer        153272
title_status     96098
transmission     95792
vin             239437
drive           192738
size            321230
type            182700
paint_color     201654
image_url        94196
description      94203
county          423857
state                0
lat              99453
long             99453
dtype: int64

แสดงเปอร์เซ็นต์ของค่า null แต่ละคอลัมน์

In [5]:
for col in df.columns:
    pct_missing = np.mean(df[col].isnull())
    print('{} - {}%'.format(col, round(pct_missing*100)))

id - 0%
url - 0%
region - 0%
region_url - 0%
price - 0%
year - 22%
manufacturer - 26%
model - 23%
condition - 58%
cylinders - 53%
fuel - 23%
odometer - 36%
title_status - 23%
transmission - 23%
vin - 56%
drive - 45%
size - 76%
type - 43%
paint_color - 48%
image_url - 22%
description - 22%
county - 100%
state - 0%
lat - 23%
long - 23%


In [6]:
NA_val = df.isna().sum()
def na_filter(na, threshold = .55): #only select variables that passees the threshold
    col_pass = []
    for i in na.keys():
        if na[i]/df.shape[0]<threshold:
            col_pass.append(i)
    return col_pass
df_cleaned = df[na_filter(NA_val)]
print(df_cleaned.columns)
print(df_cleaned.shape)

Index(['id', 'url', 'region', 'region_url', 'price', 'year', 'manufacturer',
       'model', 'cylinders', 'fuel', 'odometer', 'title_status',
       'transmission', 'drive', 'type', 'paint_color', 'image_url',
       'description', 'state', 'lat', 'long'],
      dtype='object')
(423857, 21)


จากโค้ดด้านบนทำการเลือกเอาคอลัมน์ที่มีค่าnullมากกว่า 55% ออกจากdataframe

In [7]:
df_cleaned.isnull().sum()

id                   0
url                  0
region               0
region_url           0
price                0
year             95114
manufacturer    110615
model            98473
cylinders       226178
fuel             96643
odometer        153272
title_status     96098
transmission     95792
drive           192738
type            182700
paint_color     201654
image_url        94196
description      94203
state                0
lat              99453
long             99453
dtype: int64

In [8]:
#drop row that contain null value
df_cleaned = df_cleaned.dropna(how='any', axis=0)
df_cleaned.shape

(115345, 21)

เนื่องจากข้อมูลในdataframeมีค่าnullมากเกินไปจึงทำการเอาแถวใดก็ตามที่มีค่าnullออกจากdataframe

In [9]:
df_cleaned.isnull().sum()

id              0
url             0
region          0
region_url      0
price           0
year            0
manufacturer    0
model           0
cylinders       0
fuel            0
odometer        0
title_status    0
transmission    0
drive           0
type            0
paint_color     0
image_url       0
description     0
state           0
lat             0
long            0
dtype: int64

เอา features ที่ไม่จำเป็นออก

In [10]:
df_cleaned[['id','url','region_url','image_url','region','state','description','model']]

id                                                url  \
4       7184712241  https://duluth.craigslist.org/ctd/d/rush-city-...   
5       7184711787  https://duluth.craigslist.org/ctd/d/rush-city-...   
6       7184710887  https://duluth.craigslist.org/ctd/d/rush-city-...   
7       7184710023  https://duluth.craigslist.org/ctd/d/rush-city-...   
8       7184709156  https://duluth.craigslist.org/ctd/d/rush-city-...   
...            ...                                                ...   
423849  7193376881  https://newhaven.craigslist.org/cto/d/north-br...   
423850  7195116716  https://albuquerque.craigslist.org/cto/d/albuq...   
423851  7195116656  https://albuquerque.craigslist.org/cto/d/islet...   
423852  7184919761  https://duluth.craigslist.org/cto/d/duluth-200...   
423853  7184844576  https://duluth.craigslist.org/cto/d/duluth-200...   

                                region_url  \
4            https://duluth.craigslist.org   
5            https://duluth.craigslist.org   
6            https://duluth.craigslist.org   
7            https://duluth.craigslist.org   
8            https://duluth.craigslist.org   
...                                    ...   
423849     https://newhaven.craigslist.org   
423850  https://albuquerque.craigslist.org   
423851  https://albuquerque.craigslist.org   
423852       https://duluth.craigslist.org   
423853       https://duluth.craigslist.org   

                                                image_url             region  \
4       https://images.craigslist.org/01414_g093aPtSMW...  duluth / superior   
5       https://images.craigslist.org/00W0W_io2cwEVtyh...  duluth / superior   
6       https://images.craigslist.org/00404_j5GItixnjX...  duluth / superior   
7       https://images.craigslist.org/00T0T_bwYCuKicjX...  duluth / superior   
8       https://images.craigslist.org/00O0O_j7JSL1b8Qr...  duluth / superior   
...                                                   ...                ...   
423849  https://images.craigslist.org/00N0N_cheN7HMRM5...          new haven   
423850  https://images.craigslist.org/00c0c_8Mbp45Q29F...        albuquerque   
423851  https://images.craigslist.org/00P0P_l0mIBkQXtv...        albuquerque   
423852  https://images.craigslist.org/00E0E_8o5RKLUz3o...  duluth / superior   
423853  https://images.craigslist.org/00G0G_BT0Ha3X736...  duluth / superior   

       state                                        description  \
4         mn  **Bad Credit, No Credit... No Problem!**2007 G...   
5         mn  **Bad Credit, No Credit... No Problem!**2012 F...   
6         mn  **Bad Credit, No Credit... No Problem!**2010 C...   
7         mn  **Bad Credit, No Credit... No Problem!**2011 C...   
8         mn  **Bad Credit, No Credit... No Problem!**2014 F...   
...      ...                                                ...   
423849    ct  I have posted for sale one of my company vehic...   
423850    nm  This Saturn was going to be my project car. It...   
423851    nm  NICE 2002  BLUE FORD MUSTANG GT IN EXCELENT CO...   
423852    mn  Motor runs and drives good. Transmission shift...   
423853    mn  2 owner 0 rust not from here... Leather ,roof ...   

                            model  
4           sierra classic 2500hd  
5                           f-150  
6                         equinox  
7                        traverse  
8                        explorer  
...                           ...  
423849  forester 2.5x premium awd  
423850                        sc2  
423851           mustang gt coupe  
423852                     sonata  
423853            sequoia limited  

[115345 rows x 8 columns]

In [11]:
df_cleaned.model.unique()

array(['sierra classic 2500hd', 'f-150', 'equinox', ...,
       'legacy awd premium', 'sienns xlt awd', 'juke nismo'], dtype=object)

In [12]:
uniqueValues = df_cleaned['model'].nunique()

print('Number of unique values in column "model" ')
print(uniqueValues)

Number of unique values in column "model" 
10977


In [13]:
df_final =  df_cleaned.copy().drop(['id', 'url', 'region_url', 'image_url', 'region', 'description'
                          ,'state', 'model'], axis=1)
print(df_final.shape)
print(df_final.columns)

(115345, 13)
Index(['price', 'year', 'manufacturer', 'cylinders', 'fuel', 'odometer',
       'title_status', 'transmission', 'drive', 'type', 'paint_color', 'lat',
       'long'],
      dtype='object')


เหตุผลในการเอาคอลัมน์ข้างต้นออก เพราะว่า.. <br> 
1.คอลัมน์ id เอาไว้แค่สำหรับระบุลำดับของ recordsของdataframe ซึ่งไม่มีความหมายต่อ target <br>
2.คอลัมน์ url, region_url และ image_url เป็นข้อมูลที่เป็น url ซึ่งไม่มีความหมายต่อ target <br>
3.คอลัมน์ region, state สามารถใช้ค่าของคอลัมน์ lat และ long แทนกันได้ ซึ่งให้ความหมายเดียวกันในการระบุสถานที่ <br>
4.คอลัมน์ description เป็นข้อมูลที่ไม่จำหรับการทำนายราคารถ <br>
5.คอลัมน์ model มีค่า unique value ที่มากเกินไป(4798) และข้อมูลเป็น object ซึ่งเมื่อencodeข้อมูลจะทำให้มีคอลัมน์ใหม่เพิ่มเข้ามาเป็นจำนวนมาก
และจะทำให้การสร้างโมเดลใช้เวลานานมาก <br>

In [14]:
#subsample data beacause large data requires large memory to process, and my computer has less RAM
df_final = df_final.sample(n=8000, random_state=1)
df_final.shape

(8000, 13)

In [15]:
#encode categorical variables with get_dimmies function
df_final = pd.get_dummies(df_final, drop_first=True)
print(df_final.columns)

Index(['price', 'year', 'odometer', 'lat', 'long', 'manufacturer_alfa-romeo',
       'manufacturer_aston-martin', 'manufacturer_audi', 'manufacturer_bmw',
       'manufacturer_buick', 'manufacturer_cadillac', 'manufacturer_chevrolet',
       'manufacturer_chrysler', 'manufacturer_datsun', 'manufacturer_dodge',
       'manufacturer_ferrari', 'manufacturer_fiat', 'manufacturer_ford',
       'manufacturer_gmc', 'manufacturer_harley-davidson',
       'manufacturer_honda', 'manufacturer_hyundai', 'manufacturer_infiniti',
       'manufacturer_jaguar', 'manufacturer_jeep', 'manufacturer_kia',
       'manufacturer_land rover', 'manufacturer_lexus', 'manufacturer_lincoln',
       'manufacturer_mazda', 'manufacturer_mercedes-benz',
       'manufacturer_mercury', 'manufacturer_mini', 'manufacturer_mitsubishi',
       'manufacturer_nissan', 'manufacturer_pontiac', 'manufacturer_ram',
       'manufacturer_rover', 'manufacturer_saturn', 'manufacturer_subaru',
       'manufacturer_tesla', 'manufactur

In [16]:
X = df_final.copy().drop(['price'],axis=1)
y = df_final['price']
print(X.shape)
print(y.shape)

(8000, 86)
(8000,)


เนื่องจากข้อมูลของไฟล์csvนี้มีจำนวนมากถึง 4 แสนกว่า records ซึ่งมีจำนวนมากเกินไปไม่สามารถรันบนคอมพิวเตอร์ขนาดเล็กได้ และมากเกินไปในการรันโมเดลmachine learning จึงsubsampleเหลือเพียง 8000 records 

# **Task 2 Preprocessing and Baseline Model**

In [17]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import Ridge
from sklearn.model_selection import KFold
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import RandomForestRegressor

scale = MinMaxScaler()
X_scale = scale.fit_transform(X)
kfold = KFold(n_splits=5, shuffle=True, random_state=0)
ridge = Ridge()
score = cross_val_score(ridge, X_scale, y, cv=kfold, n_jobs=-1)
print("Ridge Cross-validation scores:\n{}".format(score))
print("Ridge Mean Accuracy: {:.2f}".format(score.mean()))



Ridge Cross-validation scores:
[0.38260713 0.34404559 0.40752338 0.45708263 0.41060166]
Ridge Mean Accuracy: 0.40


In [18]:
from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly = poly.fit_transform(X_scale)
score1 = cross_val_score(ridge, X_poly, y, cv=kfold, n_jobs=-1)
print("Polynomials feature shape: {}".format(X_poly.shape))
print("Ridge Cross-validation scores:\n{}".format(score1))
print("Ridge Mean Accuracy: {:.2f}".format(score1 .mean()))

Polynomials feature shape: (8000, 3827)
Ridge Cross-validation scores:
[0.5307736  0.50308985 0.53592597 0.53066782 0.53264245]
Ridge Mean Accuracy: 0.53


ในtask 2 นี้ ได้ทำการสร้างโมเดลRidgeซึ่งเป็นlinear model โดยก่อนที่จะfitข้อมูลเข้ากับโมเดล ได้ทำการscaleข้อมูลด้วยMinMaxScaler<br>
ซึ่งจะทำให้ข้อมูลอยู่ในสเกลระหว่าง 0 ถึง 1 และจากนั้นจะวัดaccuracyด้วยการทำcross-validationซึ่งได้ความแม่นยำโดยเฉลี่ยอยู่ที่ 0.40 <br>
แล้วจากนั้นจะทำPolynomial Features เพื่อทำให้linear modelมีประสิทธิภาพมากยิ่งขึ้นเพราะการทำPolynomialจะทำให้โมเดลยืดหยุ่นและมีslopeมากขึ้นโดยใช้degree=2 แล้ววัดความแม่นยำด้วย cross-validation เหมือนเดิม จะได้ความแม่นยำเฉลี่ยอยู่ที่ 0.53 ซึ่งแม่นยำมากกว่าตอนที่ไม่ได้ทำPolynomialมาก

# **Task 4 Any model**

In [19]:
from sklearn.ensemble import RandomForestRegressor

forest = RandomForestRegressor(n_estimators=70, max_depth=14, random_state=42, n_jobs=-1)
score2 = cross_val_score(forest, X_scale, y, cv=kfold, n_jobs=-1)
print("RandomForestRegressor Cross-validation scores:\n{}".format(score2))
print("RandomForestRegressor Mean Accuracy: {:.2f}".format(score2.mean()))

RandomForestRegressor Cross-validation scores:
[0.68192184 0.63003408 0.6740909  0.69731663 0.64490941]
RandomForestRegressor Mean Accuracy: 0.67


เลือกใช้โมเดล RandomForestRegressor เพราะโมเดลพวกRandomForestเป็นโมเดลที่ไม่ต้องทำการtune parametersหรือทำpreprocessingมาก ก็สามารถทำงานและมีความแม่นยำได้ดี โดยจากโค้ดด้านบนความแม่นยำโดยเฉลี่ยอยู่ที่ 0.67

# **Task 5 Feature Selections**

In [20]:
from sklearn.feature_selection import SelectFromModel

select = SelectFromModel(RandomForestRegressor(n_estimators=70, max_depth=14, random_state=42), threshold="median")
select.fit(X_scale, y)
X_select = select.transform(X_scale)
print("X.shape: {}".format(X_scale.shape))
print("X_select.shape: {}".format(X_select.shape))



X.shape: (8000, 86)
X_select.shape: (8000, 43)


ทำการFeature Selection โดยวิธีการModel Based Feature selection ซึ่งใช้SelectFromModelโดยใช้thresholdเป็นmedian ซึ่งก็คือเลือกfeatureให้เหลือครึ่งนึง โดยmodel based feature selection จะทำการเลือกfeatureที่มีความสำคัญต่อการทำนายมากที่สุด โดยจะเลือกจากค่าfeatureที่มีค่าimportanceสูงๆ

# **Task 6 An explainable model**

In [21]:
score3 = cross_val_score(forest, X_select, y, cv=kfold, n_jobs=-1)
print("RandomForestRegressor Cross-validation scores with Feature Selection:\n{}".format(score3))
print("RandomForestRegressor Mean Accuracy with Feature Selection: {:.2f}".format(score3.mean()))

RandomForestRegressor Cross-validation scores with Feature Selection:
[0.67948178 0.6266713  0.67464978 0.69256524 0.64258055]
RandomForestRegressor Mean Accuracy with Feature Selection: 0.66


จากโค้ดด้านบนจะเห็นว่าหลังจากทำfeature selection แล้วจาก86 features เหลือแค่43features และเมื่อทำการรันโมเดลRandomForestRegressor แล้ววัดผลอีกรอบ ค่าความแม่นยำโดยเฉลี่ยอยู่ที่ 0.66 ซึ่งลดลงมาเพียงเล็กน้อยเมื่อเทียบกับการรันโดยใช้featureทั้งหมด